### PR TITLE
changed UMataData allocatorContext to OpenCLExecutionContext

### PR DESCRIPTION
**Pull request by kallaballa**
… and bind it via scope everywhere needed.

Changing it from Context to OpenCLExecutionContext poses no problem, because everywhere I need the Context I can get it from the OpenCLExecutionContext.

The pertaining issue: #20486

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Win64 OpenCL,Linux OpenCL,Linux AVX2
```

### CODE REVIEW

**Comment by kallaballa** [comment]
Fixed a problem that was preventing UMat to UMat copy *always*

**Comment by asmorkalov** [comment]
@opencv-alalek could you review?

**Comment by kallaballa** [comment]
This seems odd: https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/104719/steps/test_core/logs/tests%20summary

**Comment by opencv-alalek** [comment]
To reproduce problem build locally with `cmake ... -DWITH_OPENCL=OFF` or run with `OPENCV_OPENCL_RUNTIME=disabled`

**Comment by kallaballa** [comment]
Doc build complains about trailing white space. gonna push a fix once this build has finished

**Comment by kallaballa** [comment]
Seeing the fails and will fix.

**Comment by kallaballa** [comment]
What is the best way to reproduce those cases? is the build configuration available?

**Comment by opencv-alalek** [comment]
Failed builds are with OpenCL (default). With / without OpenCL device available (both fails).

**Comment by asmorkalov** [comment]
```
/home/ci/opencv/modules/core/src/ocl.cpp:5826:62: warning: declaration of 'pExecCtx' shadows a previous local [-Wshadow]

/home/ci/opencv/modules/core/src/ocl.cpp:5827:50: warning: declaration of 'scope' shadows a previous local [-Wshadow]

/home/ci/opencv/modules/core/src/ocl.cpp:5835:62: warning: declaration of 'pExecCtx' shadows a previous local [-Wshadow]

/home/ci/opencv/modules/core/src/ocl.cpp:5836:50: warning: declaration of 'scope' shadows a previous local [-Wshadow]
```

**Comment by kallaballa** [comment]
fixed: 784d86a

**Comment by kallaballa** [comment]
[  FAILED  ] 16 tests, listed below:
[  FAILED  ] Test_Caffe_layers.MVN/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_Caffe_layers.MVN/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_Caffe_layers.BatchNorm/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_Caffe_layers.BatchNorm/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.InstanceNorm/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_layers.InstanceNorm/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_TensorFlow_layers.batch_norm_10/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_TensorFlow_layers.batch_norm_10/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_TensorFlow_layers.batch_norm_11/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_TensorFlow_layers.batch_norm_11/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_TensorFlow_layers.batch_norm_13/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_TensorFlow_layers.batch_norm_13/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_Torch_layers.run_batch_norm/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_Torch_layers.run_batch_norm/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/1, where GetParam() = OCV/OCL_FP16

That sounds serious. Best course of action is a Window 10 VM and building with tests?

File: modules/core/src/ocl.cpp [hunk 1]
```diff
@@ -3746,6 +3746,9 @@ int Kernel::set(int i, const KernelArg& arg)
     cl_int status = 0;
     if( arg.m )
     {
+        std::shared_ptr<ocl::OpenCLExecutionContext> pExecCtx = std::static_pointer_cast<ocl::OpenCLExecutionContext>(arg.m->u->allocatorContext);
+        OpenCLExecutionContextScope scope(pExecCtx ? *pExecCtx.get() : ocl::OpenCLExecutionContext::getCurrent());
```
Comment (opencv-alalek) [inline modules/core/src/ocl.cpp:3750]: > `Kernel::`

> OpenCLExecutionContextScope

We should not change the current context selected by caller.

We need to ensure that:
- kernel is from the current (default) `OpenCLExecutionContext` (or assume that)
- and all passed buffers are compatible with the current (default) `OpenCLExecutionContext` (the same "OpenCL context")

If it is not true, then raise an error.

File: modules/core/src/umatrix.cpp [hunk 2]
```diff
@@ -1171,7 +1171,13 @@ void UMat::copyTo(OutputArray _dst) const
         if( u == dst.u && dst.offset == offset )
             return;
 
+#ifdef HAVE_OPENCL
+        std::shared_ptr<ocl::OpenCLExecutionContext> pExecCtxSrc = std::static_pointer_cast<ocl::OpenCLExecutionContext>(u->allocatorContext);
+        std::shared_ptr<ocl::OpenCLExecutionContext> pExecCtxDst = std::static_pointer_cast<ocl::OpenCLExecutionContext>(dst.u->allocatorContext);
+        if ((pExecCtxSrc && pExecCtxDst && pExecCtxSrc->getContext().ptr() == pExecCtxDst->getContext().ptr()) && u->currAllocator == dst.u->currAllocator)
```
Comment (opencv-alalek) [inline modules/core/src/umatrix.cpp:1177]: `dst` should belong to the current context.

if `src` is not related to the current context then we should force `OpenCLExecutionContextScope` below

```
Mat dst = _dst.getMat();
{
    OpenCLExecutionContextScope ...
    u->currAllocator->download(u, dst.ptr(), dims, sz, srcofs, step.p, dst.step.p);
}
```

File: modules/core/src/umatrix.cpp [hunk 3]
```diff
@@ -1212,6 +1218,9 @@ void UMat::copyTo(OutputArray _dst, InputArray _mask) const
                              ocl::memopTypeToStr(depth()), cn, mcn,
                              haveDstUninit ? " -D HAVE_DST_UNINIT" : "");
 
+        std::shared_ptr<ocl::OpenCLExecutionContext> pExecCtx = std::static_pointer_cast<ocl::OpenCLExecutionContext>(u->allocatorContext);
+        ocl::OpenCLExecutionContextScope scope(pExecCtx ? *pExecCtx.get() : ocl::OpenCLExecutionContext::getCurrent());
```
Comment (opencv-alalek) [inline modules/core/src/umatrix.cpp:1222]: > OpenCLExecutionContextScope

We should not change execution scope here.

If `src`'s scope is not the same (as dst/thread), then `.clone()` src UMat first.
If `mask`'s scope is not the same (as as dst/thread), then `.clone()` mask UMat first.

**Review by opencv-alalek — COMMENTED** [review]
Please add tests for related use cases here: https://github.com/opencv/opencv/blob/4.x/modules/core/test/test_opencl.cpp

File: modules/core/src/umatrix.cpp (hunk 3 above)
Comment (kallaballa) [reply modules/core/src/umatrix.cpp:1222]: i see!

File: modules/core/src/ocl.cpp (hunk 1 above)
Comment (kallaballa) [reply modules/core/src/ocl.cpp:3750]: alright

File: modules/core/src/umatrix.cpp (hunk 2 above)
Comment (kallaballa) [reply modules/core/src/umatrix.cpp:1177]: alright

File: modules/core/src/ocl.cpp (hunk 1 above)
Comment (kallaballa) [reply modules/core/src/ocl.cpp:3750]: How to properly compare to OpenCLExecutionContexts? I'd write the operator, but I don't know what to compare.

Comment (opencv-alalek) [reply modules/core/src/ocl.cpp:3750]: "OpenCL context" means handle of type `cl_context`.

E.g. compare these values related to `OpenCLExecutionContexts`: `execCtx.getContext().ptr()`

**Comment by asmorkalov** [comment]
@kallaballa Friendly reminder.

**Comment by asmorkalov** [comment]
@kallaballa Friendly reminder.

File: modules/core/src/ocl.cpp (hunk 1 above)
Comment (kallaballa) [reply modules/core/src/ocl.cpp]: Done

File: modules/core/src/umatrix.cpp (hunk 3 above)
Comment (kallaballa) [reply modules/core/src/umatrix.cpp]: Done, and i used the same solution for convertTo

**Comment by kallaballa** [comment]
Waiting for the tests. Apart from that i think the PR is ready.

**Comment by kallaballa** [comment]
Test tests fail due to an CV_Assert the triggers if the default context is empty. this code takes a couple of decision based on if the default context equals a UMats/buffers context. in those case if the default context is empty the UMat/buffer context should be empty too?
I'll quickly implement it like that in the hope I am right.

**Comment by kallaballa** [comment]
I thought a bit more about this problem and of course it makes sense to retain the current context and use that, because obviously the user has preferences about the context configuration or that situation wouldn't exist.
But.. working with multiple OpenCL contexts in OpenCV can become a bit tricky like multi-threaded applications with shared buffers. I such cases you sometimes have to perform quite a choreography to make sure everything is working with context at hand you or correct the situation with CPU-copies. 
There are two ideas I would like to bounce.
1. Provide a mechanism to enforce different policies for diverging context handling. e.g. The current behavior could be called "RETAIN_CONTEXT" while there could be another "AUTO_CONTEXT" that tries to make it work without error and without CPU round trip at best effort.
2. OpenCL doesn't provide neither cross context sharing nor copying. For V4D that was unacceptable because parallel workers have their own contexts and need to transfer frames to the main thread. So I implemented a workaround using CL-GL: Basically i copy the src to an OpenGL texture, blit to the dst-texture and copy the dst-texture to dst. It's a couple of GPU copies, but it's much faster than the alternative. would a PR that enables cross-context copying using CL-GL (if it is available) be interesting?

**Comment by kallaballa** [comment]
Local tests failed so i made a new patch. new local tests succeed.

**Comment by kallaballa** [comment]
I just realized i have to remove _all_ forcing of cl exec context. local test succeed.

**Comment by kallaballa** [comment]
The failing tests look suspicious and i will look into it. anyway it did some testing using my demos of V4D (on a unpublished branch) and I found several problems that prevented me from using UMat with multiple OpenCL contexts.

I will add another commit that was necessary to fix the issues i had but I am overall not sure I got this right in the first place. Maybe I misunderstood? @opencv-alalek could you have a second look on my approach to this?

Also. I currently do the following to copy cross cl-context:
```c++
CV_EXPORTS void copy_cross(const cv::UMat& src, cv::UMat& dst) {
	if(dst.empty())
		dst.create(src.size(), src.type());
	src.copyTo(dst.getMat(cv::ACCESS_READ));
}
```

Is that the right/recommended way?  I tried several other variants but to no avail.

**Comment by asmorkalov** [comment]
@kallaballa @opencv-alalek What is the PR status?

**Review by opencv-alalek — COMMENTED** [review]
Thank you for the update!

Could you prepare test cases which cover new code paths?

E.g, here: https://github.com/opencv/opencv/blob/4.x/modules/core/test/test_opencl.cpp (contains tests for `OpenCLExecutionContext`)

**Comment by kallaballa** [comment]
This is a prototype that needs rework mostly to improve code-reuse and clarity. That said it does fix the problem and considerably improves performance by avoiding errors and copies (whenever possible) on many complex workloads tested.

**Comment by kallaballa** [comment]
Also when switching between 4.x and this branch i can't find unwanted side effects.

**Comment by kallaballa** [comment]
should i pursue this one?  i know it looks very ugly but the gains are real.
}
